src/Handler/AgreementHandler.php

```diff
@@ -0,0 +1,122 @@
+use Sylius\Component\Core\Model\OrderInterface;
+use Sylius\Component\Core\Model\ShopUserInterface;
+use Webmozart\Assert\Assert;
+
+class AgreementHandler
+{
+    private AgreementHistoryRepositoryInterface $agreementHistoryRepository;
+
+    private AgreementResolverInterface $agreementResolver;
+
+    private AgreementHistoryResolverInterface $agreementHistoryResolver;
+
+    public function __construct(
+        AgreementHistoryRepositoryInterface $agreementHistoryRepository,
+        AgreementResolverInterface $agreementResolver,
+        AgreementHistoryResolverInterface $agreementHistoryResolver
+    ) {
+        $this->agreementHistoryRepository = $agreementHistoryRepository;
+        $this->agreementHistoryResolver = $agreementHistoryResolver;
+        $this->agreementResolver = $agreementResolver;
+    }
+
+    public function handleAgreements(
+        Collection $submittedAgreements,
+        string $context,
+        ?OrderInterface $order,
+        ?ShopUserInterface $shopUser
+    ): void {
+        $resolvedAgreements = $this->agreementResolver->resolve($context, []);
+
+        /** @var AgreementInterface $resolvedAgreement */
+        foreach ($resolvedAgreements as $resolvedAgreement) {
+            $submittedAgreement = $this->getSubmittedAgreement($submittedAgreements, $resolvedAgreement);
+
+            $agreementHistory = $this->setAgreementHistoryProperties($context, $order, $shopUser, $resolvedAgreement);
+
+            $resolvedAgreementHistoryState = $agreementHistory->getState();
+
+            $agreementHistoryState = $this->determineState(
+                $agreementHistory,
+                $submittedAgreement,
+                $resolvedAgreementHistoryState
+            );
+
+            if (
+                $agreementHistoryState !== $resolvedAgreementHistoryState &&
+                null !== $agreementHistory->getId()
+            ) {
+                $agreementHistory = clone $agreementHistory;
+            }
+
+            $agreementHistory->setState($agreementHistoryState);
+            $this->agreementHistoryRepository->add($agreementHistory);
+        }
+    }
+
+    private function determineState(
+        AgreementHistoryInterface $agreementHistory,
+        AgreementInterface $submittedAgreement,
+        string $resolvedAgreementHistoryState
+    ): string {
+        $agreementHistoryState = AgreementHistoryStates::STATE_SHOWN;
+
+        if (true === $submittedAgreement->isApproved()) {
+            $agreementHistoryState = AgreementHistoryStates::STATE_ACCEPTED;
+        } elseif (
+            AgreementHistoryStates::STATE_SHOWN !== $resolvedAgreementHistoryState
+            && null !== $agreementHistory->getId()
+        ) {
+            $agreementHistoryState = AgreementHistoryStates::STATE_WITHDRAWN;
+        }
+
+        return $agreementHistoryState;
+    }
+
+    private function getSubmittedAgreement(Collection $submittedAgreements, AgreementInterface $resolvedAgreement): AgreementInterface
```

Comment:
We can move it into new filter class I think

Reply:
I put the rest of the functions into other classes, I think we can leave it here, we won't test it anyway